### PR TITLE
Add clear cart and reset view

### LINKED ISSUE #40: reset to default category after payment is finished
when the payment was successful, don't just clear the cart, but also reset the view to the default category (there is a method for this, called somehow like startButtonClicked)

### PR DESCRIPTION
**Pull request by patkan**
### new functions
- cart now clearable with one click
- payup deletes cart
- categories are reset after half an hour of inactivity
- categories are reset on succesful payment

Is anything missing? This closes #40 #3 #36

### CODE REVIEW

**Comment by mgmax** [comment]
when everything is ready for merging, you could run `rebase --interactive` to merge fixup commits. The commit log is somewhat mixed up :-1: 
- [x] done

File: FabLabKasse/gui.py [hunk 1]
```diff
@@ -570,14 +584,14 @@ def askUser():
...
                 except PrinterError,  e:
-                    QtGui.QMessageBox.warning(self, "Quittung", "Drucker scheint offline zu sein."+
-                        "\nFalls du wirklich eine Quittung brauchst, melde dich bei "+
+                    QtGui.QMessageBox.warning(self, "Quittung", "Drucker scheint offline zu sein." +
+                        "\nFalls du wirklich eine Quittung brauchst, melde dich bei " +
                         "kasse@fablab.fau.de mit Datum, Uhrzeit und Betrag.")
                     logging.warning("printing receipt failed: {}".format(repr(e)))
         if paymentmethod.successful:
             paymentmethod.show_thankyou()
-            self.shoppingBackend.set_current_order(None)
```
Comment (mgmax) [inline FabLabKasse/gui.py:579]: set_current_order(None) does something else besides "clearing" the cart - it switches away from the previous, finished order!

**Comment by mgmax** [comment]
there should be some "failsafe" wrapper around libpxss, so that FabLabKasse can also be started on mac/windows.
- [x] done

File: FabLabKasse/gui.py (hunk 1 above)
Comment (patkan) [inline FabLabKasse/gui.py:579]: So the proper way would be to use the delete_current_order() method to delete the cart and then trigger an update?
Why was this function called in this place? Wouldn't deleting be the better option than "defocusing"?

File: FabLabKasse/gui.py [hunk 2]
```diff
@@ -777,15 +788,14 @@ def updateOrder(self, selectLastItem=False):
         total = self.shoppingBackend.get_current_total()
         self.summe.setText(self.shoppingBackend.format_money(self.shoppingBackend.get_current_total()))
 
-
         # disable "pay now" button on empty bill
         self.pushButton_payup.setEnabled(total > 0)
+        self.pushButton_clearCart.setEnabled(total > 0)
```
Comment (mgmax) [inline FabLabKasse/gui.py]: this is wrong: put 0 pieces of an article into the cart. Now "clear cart" is still disabled from this code.
I think setEnabled(True) would be correct.
- [x] done

File: FabLabKasse/gui.py [hunk 3]
```diff
@@ -817,8 +827,50 @@ def leaveSearch(self, keepResultsVisible=False):
...
+            reply = QtGui.QMessageBox.question(self, 'Message',
+                    u"Willst du den Warenkorb wirklich löschen?",
+                    QtGui.QMessageBox.Yes | QtGui.QMessageBox.No,
+                    QtGui.QMessageBox.No)
+            return (reply == QtGui.QMessageBox.Yes)
+        user_answer = True
+        if hide_dialog is False:
+            user_answer = ask_user()
+        if user_answer:
+            self.shoppingBackend.set_current_order(None)
```
Comment (mgmax) [inline FabLabKasse/gui.py]: this is also something else than deleting the cart, see other comment...
- [x] done

File: FabLabKasse/gui.py (hunk 1 above)
Comment (mgmax) [inline FabLabKasse/gui.py:579]: set_current_order(None): switch away from the current cart to "nothing selected". You could possibly switch back to the old order if it was not deleted.
delete_current_order(): delete the current order. Not possible if an order is already paid, as you don't want to delete it from the database. (Should imply set_current_order(None))

File: FabLabKasse/gui.py [hunk 4]
```diff
@@ -234,14 +234,22 @@ def __init__(self):
...
         self.cashPollTimer.timeout.connect(self.pollCashDevices)
         self.cashPollTimer.start()
 
-        self.pushButton_load_cart_from_app.setVisible(cfg.has_option("mobile_app","enabled") and cfg.getboolean("mobile_app","enabled"))
+        self.idleCheckTimer = QtCore.QTimer()
+        self.idleCheckTimer.setInterval(10000)
+        self.idleCheckTimer.timeout.connect(self._reset_if_idle)
+        self.idleCheckTimer.start()
+
+        self.CATEGORY_VIEW_RESET_TIME = 1800000  # idle threshold time in ms, is half an hour
```
Comment (mgmax) [inline FabLabKasse/gui.py]: as it is only used here (and referenced in a comment somewhere else), we can make this a local variable, or even better fetch the time in seconds via config.get(...,default=1800)

**Comment by patkan** [comment]
This pull request is now in #44

**Pull request closed without merging**